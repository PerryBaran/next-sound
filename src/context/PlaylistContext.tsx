'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
