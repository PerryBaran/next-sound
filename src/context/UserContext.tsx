'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
