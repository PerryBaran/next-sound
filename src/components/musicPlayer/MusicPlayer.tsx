'use client'

import { useState, useEffect, useRef } from 'react'
import css from './musicPlayer.module.css'
import Audio from './audio/Audio'
import Info from './info/Info'
import ProgresssBar from './progressbar/ProgresssBar'
import Time from './time/Time'
import MediaControls from './mediaControls/MediaControls'
import Volume from './volume/Volume'
import Playlist from './playlist/Playlist'
import { usePlaylistContext } from '@/context/PlaylistContext'

export default function MusicPlayer() {
  const { playlist, playlistIndex, skipSong, playing } = usePlaylistContext()

  const [time, setTime] = useState(0)
  const [duration, setDuration] = useState(0)
  const [volume, setVolume] = useState(0.5)
  const audioRef = useRef<HTMLAudioElement>(null)

  const song = playlist[playlistIndex] || {}

  const handleVolume = (value: number) => {
    if (value > 1) {
      setVolume(1)
    } else if (value < 0) {
      setVolume(0)
    } else {
      setVolume(value)
    }
  }

  const handleAudioTime = (value: number) => {
    if (audioRef.current) {
      audioRef.current.currentTime = value
    }
  }

  useEffect(() => {
    const updateTimer = setInterval(() => {
      if (audioRef.current) {
        const seconds = Math.round(audioRef.current.currentTime)
        setTime(seconds)
      }
    }, 250)
    return () => clearInterval(updateTimer)
  }, [time])

  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.volume = volume
    }
  }, [volume, audioRef])

  return (
    <>
      <div className={css.spacer} />
      <section className={css.container}>
        <Audio
          playing={playing}
          source={song.audio}
          audioRef={audioRef}
          setDuration={setDuration}
          skipSong={skipSong}
        />
        <Info
          albumArt={song.image}
          artistName={song.artistName}
          albumName={song.albumName}
          songName={song.songName}
        />
        <div className={css.center}>
          <ProgresssBar
            handleAudioTime={handleAudioTime}
            time={time}
            duration={duration}
          />
          <div className={css.belowBar}>
            <Time time={time} />
            <MediaControls playing={playing} skipSong={skipSong} />
            <Time time={duration} />
          </div>
        </div>
        <div className={css.centerButton}>
          <Volume volume={volume} handleVolume={handleVolume} />
        </div>
        <div className={css.centerButton}>
          <Playlist />
        </div>
      </section>
    </>
  )
}
