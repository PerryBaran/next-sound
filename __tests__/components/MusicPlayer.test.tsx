import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import { createContext, useContext } from "react"
import MusicPlayer from "../../src/components/musicPlayer/MusicPlayer"
import * as PlaylistContext from "../../src/context/PlaylistContext"

describe("Playlist", () => {
  const mockPlaylist = [
    {
      image: "/image-1",
      artistName: "artist-1",
      albumName: "album-1",
      songName: "song-1",
      audio: "audio-1",
      key: "key-1"
    },
    {
      artistName: "artist-2",
      albumName: "album-2",
      songName: "song-2",
      audio: "audio-2",
      key: "key-2"
    }
  ]
  let mockSkipSong = jest.fn()
  let mockPlay = jest.fn()

  beforeEach(() => {
    mockSkipSong = jest.fn()
    mockPlay = jest.fn()

    const mockContext = createContext({
      playlist: mockPlaylist,
      playlistIndex: 0,
      skipSong: mockSkipSong,
      handleAddToPlaylist: jest.fn(),
      removeFromPlaylist: jest.fn(),
      playing: false,
      handlePlaying: jest.fn(),
      handleShuffle: jest.fn()
    })

    jest
      .spyOn(PlaylistContext, "usePlaylistContext")
      .mockImplementation(() => useContext(mockContext))
    jest
      .spyOn(window.HTMLMediaElement.prototype, "play")
      .mockImplementation(mockPlay)
    jest
      .spyOn(window.HTMLMediaElement.prototype, "pause")
      .mockImplementation(jest.fn())
  })

  test("renders correctly", () => {
    render(<MusicPlayer />)

    expect(screen.getByTestId("audio")).toBeInTheDocument()
    expect(screen.getByAltText(/cover art/i)).toBeInTheDocument()
    expect(screen.getByDisplayValue(0)).toBeInstanceOf(HTMLInputElement)
    expect(screen.getAllByText("0:00")).toHaveLength(2)
    expect(screen.getByAltText(/skip backwards/i)).toBeInTheDocument()
    expect(screen.getByAltText(/loop/i)).toBeInTheDocument()
    expect(screen.getByAltText(/shuffle/i)).toBeInTheDocument()
    expect(screen.getByAltText(/volume/i)).toBeInTheDocument()
    expect(screen.getByAltText(/playlist/i)).toBeInTheDocument()
  })

  test("handleVolume", () => {
    render(<MusicPlayer />)
    const volumeInput = screen.getByDisplayValue(50)

    fireEvent.change(volumeInput, { target: { value: 60 } })

    expect(volumeInput).toHaveDisplayValue("60")

    fireEvent.change(volumeInput, { target: { value: 1000 } })

    expect(volumeInput).toHaveDisplayValue("100")

    fireEvent.change(volumeInput, { target: { value: -10 } })

    expect(volumeInput).toHaveDisplayValue("0")
  })

  test("handleLoop", () => {
    render(<MusicPlayer />)
    const loopButton = screen.getByAltText(/loop/i)

    expect(loopButton).toBeInTheDocument()

    fireEvent.click(loopButton)

    expect(loopButton).toHaveAttribute("alt", "loop playlist")

    fireEvent.click(loopButton)

    expect(loopButton).toHaveAttribute("alt", "loop song")
  })

  test("handleSkip", () => {
    render(<MusicPlayer />)
    const skipForwards = screen.getByAltText(/skip forwards/i)
    const loopButton = screen.getByAltText(/loop/i)

    expect(mockPlay).toBeCalledTimes(0)
    expect(mockSkipSong).toBeCalledTimes(0)

    fireEvent.click(skipForwards)

    expect(mockPlay).toBeCalledTimes(0)
    expect(mockSkipSong).toBeCalledTimes(1)
    expect(mockSkipSong).toHaveBeenLastCalledWith(true, false)

    fireEvent.click(loopButton)
    fireEvent.click(skipForwards)

    expect(mockPlay).toBeCalledTimes(0)
    expect(mockSkipSong).toBeCalledTimes(2)
    expect(mockSkipSong).toHaveBeenLastCalledWith(true, true)

    fireEvent.click(loopButton)
    fireEvent.click(skipForwards)

    expect(mockPlay).toBeCalledTimes(1)
    expect(mockSkipSong).toBeCalledTimes(2)
  })

  test("snapshot", () => {
    const { asFragment } = render(<MusicPlayer />)

    expect(asFragment()).toMatchSnapshot()
  })
})
