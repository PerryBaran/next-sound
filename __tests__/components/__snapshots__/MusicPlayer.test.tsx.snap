// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist snapshot 1`] = `
<DocumentFragment>
  <div
    class="spacer"
  />
  <section
    class="container"
  >
    <audio
      data-testid="audio"
      src="audio-1"
    />
    <div
      class="container"
    >
      <img
        alt="album-1 cover art"
        data-nimg="1"
        decoding="async"
        height="75"
        loading="lazy"
        src="/_next/image?url=%2Fimage-1&w=256&q=75"
        srcset="/_next/image?url=%2Fimage-1&w=96&q=75 1x, /_next/image?url=%2Fimage-1&w=256&q=75 2x"
        style="color: transparent;"
        width="75"
      />
      <div
        class="info"
      >
        <h3>
          artist-1
        </h3>
        <p>
          song-1
        </p>
      </div>
    </div>
    <div
      class="center"
    >
      <div
        class="container"
      >
        <input
          class="bar"
          max="0"
          min="0"
          name="time"
          type="range"
          value="0"
        />
      </div>
      <div
        class="below-bar"
      >
        <div
          class="container"
        >
          <p>
            0:00
          </p>
        </div>
        <div
          class="container"
        >
          <button>
            <img
              alt="skip backwards"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
              style="color: transparent; transform: rotate(180deg);"
              width="20"
            />
          </button>
          <button>
            <img
              alt="play"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
              style="color: transparent;"
              width="20"
            />
          </button>
          <button>
            <img
              alt="skip forwards"
              data-nimg="1"
              decoding="async"
              height="20"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
              style="color: transparent;"
              width="20"
            />
          </button>
        </div>
        <div
          class="container"
        >
          <p>
            0:00
          </p>
        </div>
      </div>
    </div>
    <div
      class="center-button loop"
    >
      <button
        class="container faded"
        type="button"
      >
        <img
          alt="loop"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </button>
    </div>
    <div
      class="center-button shuffle"
    >
      <button
        class="container faded"
        type="button"
      >
        <img
          alt="shuffle"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </button>
    </div>
    <div
      class="center-button volume"
    >
      <div
        class="container"
      >
        <input
          max="100"
          min="0"
          name="volume"
          type="range"
          value="50"
        />
        <div
          class="volume"
        >
          <button
            type="button"
          >
            <img
              alt="high volume"
              data-nimg="1"
              decoding="async"
              height="30"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
              style="color: transparent;"
              width="30"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="center-button playlist"
    >
      <button
        class="toggle-playlist faded"
        type="button"
      >
        <img
          alt="playlist"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="30"
        />
      </button>
    </div>
  </section>
</DocumentFragment>
`;
